tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -922,6 +935,289 @@ HRESULT DoSimpleReWrite(_In_ DxcLangExtensionsHelper *pHelper,
+    o.write_escaped(Filename);
+    o << '"' << '\n';
+    o.flush();
+    return lineStr;
+  }
+
+private:
+  Rewriter &TheRewriter;
+  SourceManager &SourceMgr;
+  TranslationUnitDecl *tu;
+  RewriteHelper &helper;
+  bool bNeedLineInfo;
+};
+
+HRESULT ReWriteWithLineDirective(
+    _In_ DxcLangExtensionsHelper *pHelper, _In_ LPCSTR pFileName,
+    _In_ ASTUnit::RemappedFile *pRemap, _In_ hlsl::options::DxcOpts &opts,
+    _In_ LPCSTR pDefines, std::string &warnings, std::string &result,
+    _In_opt_ dxcutil::DxcArgsFileSystem *msfPtr, IMalloc *pMalloc) {
+  raw_string_ostream o(result);
+  raw_string_ostream w(warnings);
+
+  Rewriter rewriter;
+  RewriteHelper helper;
+  {
+    // Setup a compiler instance.
+    CompilerInstance compiler;
+    std::unique_ptr<TextDiagnosticPrinter> diagPrinter =
+        llvm::make_unique<TextDiagnosticPrinter>(w,
+                                                 &compiler.getDiagnosticOpts());
+    SetupCompilerForRewrite(compiler, pHelper, pFileName, diagPrinter.get(),
+                            pRemap, opts, pDefines, msfPtr);
+
+    // Parse the source file.
+    compiler.getDiagnosticClient().BeginSourceFile(compiler.getLangOpts(),
+                                                   &compiler.getPreprocessor());
+
+    ParseAST(compiler.getSema(), false, opts.RWOpt.SkipFunctionBody);
+
+    ASTContext &C = compiler.getASTContext();
+    TranslationUnitDecl *tu = C.getTranslationUnitDecl();
+
+    if (opts.EntryPoint.empty())
+      opts.EntryPoint = "main";
+
+    rewriter.setSourceMgr(C.getSourceManager(), C.getLangOpts());
+    if (opts.RWOpt.RemoveUnusedGlobals || opts.RWOpt.RemoveUnusedFunctions) {
+      CollectRewriteHelper(tu, opts.EntryPoint.data(), helper,
+                           opts.RWOpt.RemoveUnusedGlobals,
+                           opts.RWOpt.RemoveUnusedFunctions, w);
+      RewriteVisitor visitor(rewriter, tu, helper);
+      visitor.TraverseDecl(tu);
+    }
+    // TODO: support ExtractEntryUniforms, GlobalExternByDefault, SkipStatic, SkipFunctionBody.
+    if (compiler.getDiagnosticClient().getNumErrors() > 0) {
+      o.flush();
+      w.flush();
+      return E_FAIL;
+    }
+
+    helper.macros = CollectSemanticDefinesParsedByCompiler(compiler, pHelper);
+
+    if (opts.RWOpt.KeepUserMacro)
+      helper.userMacros = CollectUserMacrosParsedByCompiler(compiler);
+
+  }
+
+  {
+    CComPtr<AbstractMemoryStream> pOutputStream;
+    IFT(CreateMemoryStream(pMalloc, &pOutputStream));
+
+    raw_stream_ostream outStream(pOutputStream.p);
+
+    IFT(msfPtr->RegisterOutputStream(L"output.bc", pOutputStream));
+
+    llvm::MemoryBuffer *pMemBuf = pRemap->second;
+    std::unique_ptr<llvm::MemoryBuffer> pBuffer(
+        llvm::MemoryBuffer::getMemBufferCopy(pMemBuf->getBuffer(), pFileName));
+
+    std::unique_ptr<ASTUnit::RemappedFile> pRemap(
+        new ASTUnit::RemappedFile(pFileName, pBuffer.release()));
+
+    CompilerInstance compiler;
```

Comment:
I think I need an explanation of why this second compiler instance and the preprocessor run is needed.